engine/flink/model-util/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/OuterJoinTransformerSpec.scala

```diff
@@ -82,74 +82,59 @@ class OuterJoinTransformerSpec extends FunSuite with FlinkSpec with Matchers {
     ))
 
     val key = "fooKey"
-    val input1 = List(
-      OneRecord(key, 0, -1),
-      OneRecord(key, 2, -1)
-    )
+    val input1 = new BlockingQueueSource[OneRecord](OneRecord.timestampExtractor)
     val input2 = List(
       OneRecord(key, 1, 123)
     )
 
-    val results = runProcess(process, input1, input2)
+    val collectingListener = ResultsCollectingListenerHolder.registerRun(identity)
+    val (id, stoppableEnv) = runProcess(process, input1, input2, collectingListener)
+
+    input1.add(OneRecord(key, 0, -1))
+    // We can't be sure that main records will be consumed after matching joined records so we need to wait for them.
+    Eventually.eventually {
+      OuterJoinTransformerSpec.elementsAddedToState.size() == input2.size
+    }
+    input1.add(OneRecord(key, 2, -1))
+    input1.finish()
+
+    stoppableEnv.waitForJobState(id.getJobID, process.id, ExecutionState.FINISHED)()
 
-    val outValues = results.nodeResults(EndNodeId)
+    val outValues = collectingListener.results[Any].nodeResults(EndNodeId)
       .filter(_.variableTyped(KeyVariableName).contains(key))
       .map(_.variableTyped[java.lang.Integer](OutVariableName).get)
 
     outValues shouldEqual List(null, 123)
   }
 
-  private def runProcess(testProcess: EspProcess, input1: List[OneRecord], input2: List[OneRecord]): TestResults[Any] = {
-    val collectingListener = ResultsCollectingListenerHolder.registerRun(identity)
+  private def runProcess(testProcess: EspProcess, input1: BlockingQueueSource[OneRecord], input2: List[OneRecord], collectingListener: ResultsCollectingListener) = {
     val model = modelData(input1, input2, collectingListener)
     val stoppableEnv = flinkMiniCluster.createExecutionEnvironment()
     val registrar = FlinkStreamingProcessRegistrar(new FlinkProcessCompiler(model), model.processConfig, ExecutionConfigPreparer.unOptimizedChain(model, None))
     registrar.register(new StreamExecutionEnvironment(stoppableEnv), testProcess, ProcessVersion.empty, Some(collectingListener.runId))
-    stoppableEnv.executeAndWaitForFinished(testProcess.id)()
-    collectingListener.results[Any]
+    val id = stoppableEnv.executeAndWaitForStart(testProcess.id)
+    (id, stoppableEnv)
   }
 
-  private def modelData(input1: List[OneRecord], input2: List[OneRecord], collectingListener: ResultsCollectingListener) =
+  private def modelData(input1: BlockingQueueSource[OneRecord], input2: List[OneRecord], collectingListener: ResultsCollectingListener) =
     LocalModelData(ConfigFactory.empty(), new OuterJoinTransformerSpec.Creator(input1, input2, collectingListener))
 
 }
 
 object OuterJoinTransformerSpec {
 
-  class Creator(mainRecords: List[OneRecord], joinedRecords: List[OneRecord], collectingListener: ResultsCollectingListener) extends EmptyProcessConfigCreator {
+  val elementsAddedToState = new ConcurrentLinkedQueue[StringKeyedValue[AnyRef]]()
```

Comment:
I wonder if you can use collectingListener for that and get rid of CoProcessFunctionInterceptor altogether? 
Although TBH I don't remember how it deals with joins ;)